ait/core/server/server.py

```diff
@@ -0,0 +1,255 @@
+                           self.broker.plugins +
+                           self.broker.inbound_streams +
+                           self.broker.outbound_streams)
+
+    def wait(self):
+        """
+        Starts all greenlets for concurrent processing.
+        Joins over all greenlets that are not servers.
+        """
+        for greenlet in (self.greenlets + self.servers):
+            log.info("Starting {} greenlet...".format(greenlet))
+            greenlet.start()
+
+        gevent.joinall(self.greenlets)
+
+    def _load_streams(self):
+        """
+        Reads, parses and creates streams specified in config.yaml.
+        """
+        common_err_msg = 'No valid {} telemetry stream configurations found. '
+        specific_err_msg = {'inbound': 'No telemetry will be received (or displayed).',
+                            'outbound': 'No telemetry will be published.'}
+        err_msgs = {}
+
+        for stream_type in ['inbound', 'outbound']:
+            err_msgs[stream_type] = common_err_msg.format(stream_type) + specific_err_msg[stream_type]
+            streams = ait.config.get('server.{}-streams'.format(stream_type))
+
+            if streams is None:
+                log.warn(err_msgs[stream_type])
+            else:
+                for index, s in enumerate(streams):
+                    try:
+                        strm = self._create_stream(s['stream'], stream_type)
+                        if stream_type == 'inbound' and type(strm) == PortInputStream:
+                            self.servers.append(strm)
+                        elif stream_type == 'inbound':
+                            self.inbound_streams.append(strm)
+                        elif stream_type == 'outbound':
+                            self.outbound_streams.append(strm)
+                        log.info('Added {} stream {}'.format(stream_type, strm))
+                    except Exception:
+                        exc_type, value, tb = sys.exc_info()
+                        log.error('{} creating {} stream {}: {}'.format(exc_type,
+                                                                        stream_type,
+                                                                        index,
+                                                                        value))
+        if not self.inbound_streams and not self.servers:
+            log.warn(err_msgs['inbound'])
+
+        if not self.outbound_streams:
+            log.warn(err_msgs['outbound'])
+
+    def _create_stream(self, config, stream_type):
+        """
+        Creates a stream from its config.
+
+        Params:
+            config:       stream configuration as read by ait.config
+            stream_type:  either 'inbound' or 'outbound'
+        Returns:
+            stream:       a Stream
+        Raises:
+            ValueError:   if any of the required config values are missing
+        """
+        if stream_type not in ['inbound', 'outbound']:
```

Comment:
Is this check necessary? This is a "private" function and it can only ever get one of these two values. Super nit-picky and not the end of the world, but a bit of a code smell to me.

Reply:
Split `create_stream` into 2 separate functions for inbound and outbound stream types